compiler/rustc_data_structures/src/sharded.rs

```diff
@@ -1,31 +1,25 @@
 use crate::fx::{FxHashMap, FxHasher};
 #[cfg(parallel_compiler)]
-use crate::sync::is_dyn_thread_safe;
-use crate::sync::{CacheAligned, Lock, LockGuard};
+use crate::sync::{is_dyn_thread_safe, CacheAligned};
+use crate::sync::{Lock, LockGuard};
 use std::borrow::Borrow;
 use std::collections::hash_map::RawEntryMut;
 use std::hash::{Hash, Hasher};
 use std::mem;
 
-#[cfg(parallel_compiler)]
 // 32 shards is sufficient to reduce contention on an 8-core Ryzen 7 1700,
 // but this should be tested on higher core count CPUs. How the `Sharded` type gets used
 // may also affect the ideal number of shards.
-const SHARD_BITS: usize = 5;
-
-#[cfg(not(parallel_compiler))]
-const SHARD_BITS: usize = 0;
+const SHARD_BITS: usize = if cfg!(parallel_compiler) { 5 } else { 0 };
 
 pub const SHARDS: usize = 1 << SHARD_BITS;
 
 /// An array of cache-line aligned inner locked structures with convenience methods.
-pub struct Sharded<T> {
-    /// This mask is used to ensure that accesses are inbounds of `shards`.
-    /// When dynamic thread safety is off, this field is set to 0 causing only
-    /// a single shard to be used for greater cache efficiency.
+/// A single field is used when the compiler uses only one thread.
+pub enum Sharded<T> {
+    Single(Lock<T>),
```

Comment:
We can use `RefCell` directly if it brought visible performeance benefit. Then we need to add `Sharded` to `marker.rs`

Reply:
That doesn't work as we need to return a single type for both cases. With https://github.com/rust-lang/rust/pull/111713 landed however we can add a lock operation to `Sharded` that combines the branches on `is_dyn_thread_safe`.

Reply:
Yea I mean `Single(RefCell<T>)`

Reply:
Oh `fn lock_shards` requires the return type `Lock<T>`. Just forget it :)